release/Earthfile

```diff
@@ -63,18 +124,43 @@ release-github:
     RUN file ./release/earthly-linux-arm64 | grep "ELF 64-bit"
     RUN file ./release/earthly-windows-amd64.exe | grep "PE32"
     ARG BODY="No details provided"
-    RUN --secret GITHUB_TOKEN=+secrets/earthly-technologies/github/griswoldthecat/token test -n "$GITHUB_TOKEN"
+    ARG GITHUB_SECRET_PATH="+secrets/earthly-technologies/github/griswoldthecat/token"
+    RUN --secret GITHUB_TOKEN="$GITHUB_SECRET_PATH" test -n "$GITHUB_TOKEN" && echo $GITHUB_TOKEN
     RUN --push \
-        --secret GITHUB_TOKEN=+secrets/earthly-technologies/github/griswoldthecat/token \
+        --secret GITHUB_TOKEN="$GITHUB_SECRET_PATH" \
+        set -e; \
+        # test github token works
+        curl -H "Authorization: token $GITHUB_TOKEN" "https://api.github.com/user" | jq .login | grep "$GITHUB_USER"; \
+        # first delete any previously released files (needed in case the previous upload attempt failed)
+        PREV_RELEASE_ID=$(curl -H "Authorization: token $GITHUB_TOKEN" "https://api.github.com/repos/$GITHUB_USER/$EARTHLY_REPO/releases/tags/$RELEASE_TAG" | jq .id); \
+        if [ -n "$PREV_RELEASE_ID" ] && [ "$PREV_RELEASE_ID" != "null" ]; then \
+          echo "deleting previous release ID $PREV_RELEASE_ID"; \
+          curl -H "Authorization: token $GITHUB_TOKEN" --request DELETE "https://api.github.com/repos/$GITHUB_USER/$EARTHLY_REPO/releases/$PREV_RELEASE_ID"; \
+        fi; \
+        # next make sure any previous tag is deleted (otherwise github will create an untagged release name vX.Y.Z)
+        curl -H "Authorization: token $GITHUB_TOKEN" --request DELETE "https://api.github.com/repos/$GITHUB_USER/$EARTHLY_REPO/git/refs/tags/$RELEASE_TAG"; \
+        # next, upload binaries
         github-release upload \
-        --owner earthly \
-        --repo earthly \
-        --prerelease true \
+        --owner "$GITHUB_USER" \
+        --repo "$EARTHLY_REPO" \
+        --prerelease false \
```

Comment:
Is this right? We should first write the release notes and only afterward uncheck the prerelease flag.

Reply:
If we leave this as true, then we must manually go in and uncheck the pre-release flag between the github release and the homebrew release.

I was thinking that we could write the release notes in parallel, and once the release shows up in github we can edit it there. The downside with this is it will introduce a race-condition of sorts where users could download a version of earthly that does not contain release notes.

Another idea is we could start committing release notes into a changelog in git, and this release script could pull out the appropriate details for the release we are doing. If the changelog notes don't exist, then the release would exit out. Maintaining a changelog would give users a single file to look over in case they wanted to know what feature introduced a change that may have created (or fixed) a bug.

Reply:
we now have a `CHANGELOG.md` and corresponding `changelogparser.py` which passes this as the body for the release text.